File: application-ver2/app/src/main/java/com/example/heartbeat/PPG/FragmentPPG.java
package com.example.heartbeat.PPG;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.FrameLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.example.heartbeat.Command;
import com.example.heartbeat.MenuActivity;
import com.example.heartbeat.R;

public class FragmentPPG extends Fragment{
    Button start;
    Button pause;
    Command MyCmd;
    FrameLayout frameLayout;
    TextView sensorField;
    View view;
    RealTimeGraphPPG myGraph;
    Thread realTimeThread;
    Activity activity;
    Boolean threadFlag;
    Boolean initialStartFlag;
    //ppg의 경우, 초기에 일정 인터벌 쉬었다가 그래프에 그려줘야 하는데, runOnUiThread를 우아하게 멈출 방법이 아직 생각이 안나서 임시방편으로 flag와 if문으로 인터벌 쉬도록 구현함

    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        view = inflater.inflate(R.layout.fragmentppg, container, false);

        start = (Button)view.findViewById(R.id.start);
        pause = (Button)view.findViewById(R.id.pause);

        sensorField = (TextView)view.findViewById(R.id.temp_sensor_value);
        myGraph = new RealTimeGraphPPG(view);
        realTimeThread = null;
        threadFlag = false;

        start.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v){
                ((MenuActivity)getActivity()).setViewField(view, "ppg");
                ((MenuActivity)getActivity()).sendStrCmd(MyCmd.str_readppg0);

                start.setVisibility(View.GONE);
                pause.setVisibility(View.VISIBLE);
                threadFlag = true;
                initialStartFlag = true;
                realTimeStart();


            }
        });

        pause.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v){
                ((MenuActivity)getActivity()).setViewField(view, "stop");
                ((MenuActivity)getActivity()).sendStrCmd(MyCmd.str_stop);
                start.setVisibility(View.VISIBLE);
                pause.setVisibility(View.GONE);
                threadFlag = false;
            }
        });

        LayoutInflater layoutInflater = (LayoutInflater)getActivity().getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        frameLayout = ((MenuActivity)getActivity()).getFrameLayout();

        MyCmd = new Command();

        return view;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

        if (context instanceof Activity)
            activity = (Activity) context;
    }
    @Override
    public void onDetach() {
        ((MenuActivity)getActivity()).sendStrCmd(MyCmd.str_stop);
        ((MenuActivity)getActivity()).mode="stop";
        frameLayout.setVisibility(View.GONE);
        super.onDetach();

    }


    public void realTimeStart(){
        realTimeThread = new Thread(new Runnable() {
            
            @Override
            public void run() {
                while(threadFlag){
                    try {
                        if(initialStartFlag){
                            Thread.sleep(10000);
                            initialStartFlag = false;
                        }
                        Thread.sleep(100);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                    activity.runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                                myGraph.addEntry( ((MenuActivity) activity).ppgData.getGrnCnt(), ((MenuActivity) activity).ppgData.getGrn2Cnt());
                        }
                    });
                }
            }
        });

        realTimeThread.start();
    }


}

File: Application_part/project_kotlin_version/Kotlin_ver/app/src/main/java/com/example/kotlin_ver/PPG/FragmentPPG.kt
package com.example.kotlin_ver.PPG

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.FrameLayout
import android.widget.TextView
import androidx.fragment.app.Fragment
import com.example.kotlin_ver.Command
import com.example.kotlin_ver.MenuActivity
import com.example.kotlin_ver.R


class FragmentPPG : Fragment() {
    lateinit var start: Button
    lateinit var pause: Button
    lateinit var ppgValueField: TextView
    lateinit var confidenceField: TextView
    lateinit var activityStateField: TextView
    lateinit var myGraph: RealTimeGraphPPG

    var activity: Activity? = null

    val ppgDataList: ArrayList<PPGData> = ArrayList()

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val view = inflater.inflate(R.layout.fragmentppg, container, false)
        start = view.findViewById(R.id.start)
        pause = view.findViewById(R.id.pause)
        ppgValueField = view.findViewById(R.id.ppg_sensor_value)
        confidenceField = view.findViewById(R.id.ppg_sensor_confidence_value)
        activityStateField = view.findViewById(R.id.ppg_sensor_activity_value)

        myGraph = RealTimeGraphPPG(view)

        start.setOnClickListener {
            (activity as MenuActivity).sendStrCmd(Command.str_readppg0)
            (activity as MenuActivity).mode = "PPG"
            start.visibility = View.GONE
            pause.visibility = View.VISIBLE
        }
        pause.setOnClickListener {
            (activity as MenuActivity).sendStrCmd(Command.str_stop)
            (activity as MenuActivity).mode = "stop"
            start.visibility = View.VISIBLE
            pause.visibility = View.GONE
        }

        return view
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is Activity) activity = context
    }

    override fun onDetach() {
        (activity as MenuActivity).sendStrCmd(Command.str_stop)
        (getActivity() as MenuActivity).mode = "stop"
        (getActivity() as MenuActivity).fragmentContainerLayout.visibility = View.GONE
        super.onDetach()
    }

    fun addData(hexaData: ByteArray){
        val ppg = PPGData(hexaData)
        ppgDataList.add(ppg)

        ppgValueField.text = "${ppg.heartRate}"
        confidenceField.text = "${ppg.heartRateConfidence}"
        activityStateField.text = "${ppg.ppgActivityString}"

        activity!!.runOnUiThread {
            myGraph.addEntry(ppg)
        }
    }


}